File: app/src/main/java/com/pleon/buyt/database/repository/StatisticsRepository.java
package com.pleon.buyt.database.repository;

import android.app.Application;
import android.os.AsyncTask;

import com.pleon.buyt.database.AppDatabase;
import com.pleon.buyt.database.SingleLiveEvent;
import com.pleon.buyt.database.dao.PurchaseDao;
import com.pleon.buyt.model.Category;
import com.pleon.buyt.model.Statistics;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

public class StatisticsRepository { // TODO: make this class singleton

    private PurchaseDao mPurchaseDao;
    private SingleLiveEvent<Statistics> averagePurchaseCost;

    public StatisticsRepository(Application application) {
        mPurchaseDao = AppDatabase.getDatabase(application).purchaseDao();
        averagePurchaseCost = new SingleLiveEvent<>();
    }

    public LiveData<Statistics> getStatistics(int period, @Nullable Category filter) {
        new GetStatisticsTask(mPurchaseDao, averagePurchaseCost, period, filter).execute();
        return averagePurchaseCost;
    }

    private static class GetStatisticsTask extends AsyncTask<Void, Void, Statistics> {

        private PurchaseDao purchaseDao;
        private MutableLiveData<Statistics> statistics;
        private int period;
        private Category filter;

        GetStatisticsTask(PurchaseDao purchaseDao, MutableLiveData<Statistics> statistics,
                          int period, @Nullable Category filter) {
            this.purchaseDao = purchaseDao;
            this.statistics = statistics;
            this.period = period;
            this.filter = filter;
        }

        @Override
        protected Statistics doInBackground(Void... voids) {
            return purchaseDao.getStats(period, filter);
        }

        @Override
        protected void onPostExecute(Statistics statistics) {
            this.statistics.setValue(statistics);
        }
    }
}

File: app/src/main/java/com/pleon/buyt/database/repository/StatisticsRepository.kt
package com.pleon.buyt.database.repository

import android.app.Application
import androidx.lifecycle.LiveData
import com.pleon.buyt.database.AppDatabase
import com.pleon.buyt.database.SingleLiveEvent
import com.pleon.buyt.model.Category
import com.pleon.buyt.model.Statistics
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread

class StatisticsRepository(application: Application) { // TODO: make this class singleton

    private val purchaseDao = AppDatabase.getDatabase(application)!!.purchaseDao()
    private val statistics = SingleLiveEvent<Statistics>()

    fun getStatistics(period: Int, filter: Category?): LiveData<Statistics> {
        doAsync {
            val states = purchaseDao.getStatistics(period, filter)
            uiThread { statistics.value = states }
        }
        return statistics
    }
}
